R/writeMSData.R

```diff
@@ -0,0 +1,186 @@
+#'     `"totIonCurrent"`, `"basePeakMZ"` and `"basePeakIntensity"` on the data.
+#'
+#' @param x `OnDiskMSnExp` or `MSnExp` object.
+#'
+#' @param files `character` with the file name(s). Its length has to match the
+#'     number of samples/files of `x`.
+#'
+#' @param outformat `character(1)` defining the format of the output files.
+#'
+#' @param verbose `logical(1)` if progress messages should be displayed.
+#'
+#' @param copy `logical(1)` if metadata (data processings, original file names
+#'     etc) should be copied from the original files.
+#' 
+#' @param software_processing optionally provide specific data processing steps.
+#'     See documentation of the `software_processing` parameter of
+#'     [mzR::writeMSData()].
+#'
+#' @author Johannes Rainer
+#'
+#' @md
+#'
+#' @noRd
+.writeMSData <- function(x, files, outformat = c("mzml", "mzxml"),
+                         verbose = isMSnbaseVerbose(),
+                         copy = TRUE,
+                         software_processing = NULL) {
+    ## Check input.
+    outformat <- match.arg(outformat)
+    if (missing(files))
+        stop("'files' is required!")
+    if (length(files) != length(fileNames(x)))
+        stop("length of 'files' has to match the number of samples")
+    if (verbose)
+        message("Writing ", length(files), " ", outformat, " files.")
+    ## Split per file.
+    x_split <- splitByFile(x, f = factor(fileNames(x)))
+    ## Using mapply below - in principle we could then even switch to
+    ## bpmapply to perform parallel saving of files.
+    res <- mapply(x_split, files, FUN = function(z, f, outformat,
+                                                 software_processing,
+                                                 verbose){
+        if (verbose)
+            message("Saving file ", basename(f), "...", appendLF = FALSE)
+        if (file.exists(f))
+            stop("File ", f, " does already exist! Please use a different ",
+                 "file name.")
+        ## o get all peak data. Calling spectrapply will apply also all lazy
+        ##   processing steps. We're not calling the `as` method to avoid R
+        ##   having to look through all namespaces to find the function.
+        pks <- spectrapply(z, FUN = function(sp) cbind(sp@mz, sp@intensity),
+                           BPPARAM = SerialParam())
+        if (is(x, "OnDiskMSnExp")) {
+            hdr <- fData(z)
+        } else {
+            ## MSnExp: feature data does not provide all the data we need.
+            hdr <- data.frame(do.call(rbind,
+                                      spectrapply(z, FUN = .spectrum_header,
+                                                  BPPARAM = SerialParam())),
+                              check.names = FALSE)
+        }
+        ## o Re-calculate stuff we don't have or which might have been
+        ##   changed:
+        new_vals <- do.call(rbind, lapply(pks, function(sp) {
```

Comment:
You are creating exact these values in `.spectrum_header` already. How could they change from line 64 to line 71 without being manipulated by us? If only the `fData` values could have been modified this `do.call` and the following `hdr$...` should go into the `if(is(x, "OnDiskMSnExp"))` block above (line 61).

Reply:
You see everything ;) For `MSnExp` that's correct, for `OnDiskMSnExp` not, because the `fData` for these objects is in fact the `header` at the time when the file is read. So it's better/saver to (re)calculate for `OnDiskMSnExp`.
Yes, I'll put that into the `if(is...`. :+1: